### PR TITLE
Tie up loose ends around making check(cleanup =) defunct

### LINKED ISSUE #2448: Alternative to `cleanup = FALSE`
It seems like argument `cleanup` of `devtools::check()` has been defunct in **devtools** v2.4.4. I realized this after updating **devtools** to v2.4.4 and then running `R CMD check` via RStudio's keyboard shortcut `Ctrl + Shift + E` on one of my packages. I have deselected option "Cleanup output after successful R CMD check" in RStudio's global options. My question is now what the recommended alternative to `cleanup = FALSE` is. I couldn't find anything about that in the documentation and neither by searching the internet.

### PR DESCRIPTION
**Pull request by jennybc**
Closes #2448

### CODE REVIEW

File: R/check.R [hunk 1]
```diff
@@ -39,17 +39,18 @@
 #'
 #' @return An object containing errors, warnings, and notes.
 #' @template devtools
+#' @inheritParams rcmdcheck::rcmdcheck
```
Comment (jennybc) [inline R/check.R:42]: Let's just let rcmdcheck explain what `check_dir` does.

File: R/check.R [hunk 2]
```diff
@@ -39,17 +39,18 @@
...
 #' @param build_args Additional arguments passed to `R CMD build`
-#' @param check_dir the directory in which the package is checked
-#'   compatibility. `args = "--output=/foo/bar"` can be used to change the
-#'   check directory.
 #' @param ... Additional arguments passed on to [pkgbuild::build()].
 #' @param vignettes If `FALSE`, do not build or check vignettes, equivalent to
 #'   using `args = '--ignore-vignettes' and `build_args = '--no-build-vignettes'.
-#' @param cleanup Deprecated.
+#' @param cleanup `r lifecycle::badge("deprecated")` `check()` no longer does
+#'   any cleanup, so if you want to inspect the built package, store the return
```
Comment (jennybc) [inline R/check.R:51]: It seems worth mentioning that you should store the return value of `check()`, so that garbage collection doesn't happen right away.

File: R/check.R (hunk 1 above)
Comment (fweber144) [reply R/check.R:42]: Shouldn't this read "if this is `NULL`" (i.e., without the "not")? Would be an rcmdcheck issue, though. Also, with `devtools::check()`, I couldn't reproduce this deletion of the temporary directory upon calling `gc()` (and storing the `devtools::check()` return value beforehand). But I might just not have understood this correctly or this might be something that only works in the context of rcmdcheck, but not `devtools::check()` (in which case the inheritance of this documentation would be inappropriate, though).

Comment (jennybc) [reply R/check.R:43]: Yeah I think you are right about the re-wording re: (not) `NULL`, but also, yes, that's a change that needs to happen in rcmdcheck (would you like to make a PR there?). I think it's clear that devtools should transmit rcmdcheck's docs for `check_dir`, regardless.

File: R/check.R [hunk 3]
```diff
@@ -159,22 +160,19 @@ can_document <- function(pkg) {
...
-#'   (`--no-manual`).
+#' @param cran if `TRUE` (the default), check using the same settings as CRAN
+#'   uses.
+#' @param remote Sets `_R_CHECK_CRAN_INCOMING_REMOTE_` env var. If `TRUE`,
+#'   performs a number of CRAN incoming checks that require remote access.
+#' @param incoming Sets `_R_CHECK_CRAN_INCOMING_` env var. If `TRUE`, performs a
+#'   number of CRAN incoming checks.
+#' @param force_suggests Sets `_R_CHECK_FORCE_SUGGESTS_`. If `FALSE` (the
+#'   default), check will proceed even if all suggested packages aren't found.
+#' @param run_dont_test Sets `--run-donttest` so that examples surrounded in
```
Comment (jennybc) [inline R/check.R:171]: Also fixing "tests" --> "examples" while we're here.

File: R/check.R (hunk 2 above)
Comment (fweber144) [reply R/check.R:52]: In my experiments (see enumeration point 1 in https://github.com/r-lib/devtools/issues/2448#issuecomment-1206127954), I didn't have to store the return value of `check()` to ensure that the `.Rcheck` directory is not deleted.

File: R/check.R (hunk 1 above)
Comment (fweber144) [reply R/check.R:43]: Sorry, my [comment](https://github.com/r-lib/devtools/pull/2455#discussion_r940886421) was incorrect:

> upon calling `gc()` (and storing the `devtools::check()` return value beforehand).

should have read

> upon calling `gc()` (and not storing the `devtools::check()` return value beforehand)

(note the "not").

Comment (fweber144) [reply R/check.R:43]: Oh, the reason for why the deletion upon garbage collection doesn't happen in `devtools::check()` might be that it uses `check_dir = tempdir()` by default.

Comment (fweber144) [reply R/check.R:43]: So not `check_dir = NULL`.

Comment (fweber144) [reply R/check.R:43]: Yes, I can confirm that when using `devtools::check(document = FALSE, check_dir = NULL)` in my reprex from #2448, then the `.Rcheck` directory (contained in a temporary folder) is deleted afterwards.

Comment (fweber144) [reply R/check.R:43]: In contrast, `devtools::check(document = FALSE)` keeps the `.Rcheck` directory.

Comment (fweber144) [reply R/check.R:43]: Assigning the `devtools::check()` return value to some object doesn't change anything in these two observations (https://github.com/r-lib/devtools/pull/2455#discussion_r940908360 and https://github.com/r-lib/devtools/pull/2455#discussion_r940908806).

Comment (fweber144) [reply R/check.R:43]: Now I have to see how this behaves in rcmdcheck. Afterwards, I can submit a PR.

Comment (fweber144) [reply R/check.R:43]: Assigning the return value to some objects also doesn't seem to be necessary in rcmdcheck:
```r
rcmdcheck::rcmdcheck()
## --> `.Rcheck` directory deleted afterwards.
out <- rcmdcheck::rcmdcheck()
## --> `.Rcheck` directory deleted afterwards.
rcmdcheck::rcmdcheck(check_dir = NULL)
## --> `.Rcheck` directory deleted afterwards.
out <- rcmdcheck::rcmdcheck(check_dir = NULL)
## --> `.Rcheck` directory deleted afterwards.
rcmdcheck::rcmdcheck(check_dir = tempdir())
## --> `.Rcheck` directory *not* deleted afterwards, not even after calling gc().
out <- rcmdcheck::rcmdcheck(check_dir = tempdir())
## --> `.Rcheck` directory *not* deleted afterwards(, not even after calling gc()).
```
I'll adapt the rcmdcheck PR (that you requested [above](https://github.com/r-lib/devtools/pull/2455#discussion_r940888612)) accordingly.

File: R/check.R (hunk 2 above)
Comment (fweber144) [reply R/check.R:52]: I think the part about storing the return value can be omitted from the docs, given https://github.com/r-lib/devtools/pull/2455#discussion_r940911269.

File: R/check.R (hunk 1 above)
Comment (hadley) [reply R/check.R:43]: I can't reproduce that behaviour:

``` r
out <- rcmdcheck::rcmdcheck()
#> ── R CMD build ─────────────────────────────────────────────────────────────────
#> * checking for file ‘.../DESCRIPTION’ ... OK
#> * preparing ‘waldo’:
#> * checking DESCRIPTION meta-information ... OK
#> * checking for LF line-endings in source and make files and shell scripts
#> * checking for empty or unneeded directories
#> * building ‘waldo_0.4.0.9000.tar.gz’
#> 
#> ── R CMD check ─────────────────────────────────────────────────────────────────
#> * using log directory ‘/private/tmp/RtmpK2EwwH/filed4606dfeb7b2/waldo.Rcheck’
#> * using R version 4.1.2 Patched (2022-02-15 r81747)
#> * using platform: x86_64-apple-darwin17.0 (64-bit)
#> * using session charset: UTF-8
#> * checking for file ‘waldo/DESCRIPTION’ ... OK
#> * this is package ‘waldo’ version ‘0.4.0.9000’
#> * package encoding: UTF-8
#> * checking package namespace information ... OK
#> * checking package dependencies ... OK
#> * checking if this is a source package ... OK
#> * checking if there is a namespace ... OK
#> * checking for executable files ... OK
#> * checking for hidden files and directories ... OK
#> * checking for portable file names ... OK
#> * checking for sufficient/correct file permissions ... OK
#> * checking whether package ‘waldo’ can be installed ... OK
#> * checking installed package size ... OK
#> * checking package directory ... OK
#> * checking DESCRIPTION meta-information ... OK
#> * checking top-level files ... OK
#> * checking for left-over files ... OK
#> * checking index information ... OK
#> * checking package subdirectories ... OK
#> * checking R files for non-ASCII characters ... OK
#> * checking R files for syntax errors ... OK
#> * checking whether the package can be loaded ... OK
#> * checking whether the package can be loaded with stated dependencies ... OK
#> * checking whether the package can be unloaded cleanly ... OK
#> * checking whether the namespace can be loaded with stated dependencies ... OK
#> * checking whether the namespace can be unloaded cleanly ... OK
#> * checking dependencies in R code ... OK
#> * checking S3 generic/method consistency ... OK
#> * checking replacement functions ... OK
#> * checking foreign function calls ... OK
#> * checking R code for possible problems ... OK
#> * checking Rd files ... OK
#> * checking Rd metadata ... OK
#> * checking Rd cross-references ... OK
#> * checking for missing documentation entries ... OK
#> * checking for code/documentation mismatches ... OK
#> * checking Rd \usage sections ... OK
#> * checking Rd contents ... OK
#> * checking for unstated dependencies in examples ... OK
#> * checking examples ... OK
#> * checking for unstated dependencies in ‘tests’ ... OK
#> * checking tests ...
#>   Running ‘testthat.R’
#>  OK
#> * checking PDF version of manual ... OK
#> * DONE
#> 
#> Status: OK
path <- out$checkdir
file.exists(path)
#> [1] TRUE
rm(out); gc()
#>           used (Mb) gc trigger (Mb) limit (Mb) max used (Mb)
#> Ncells  822989 44.0    1314504 70.3         NA  1314504 70.3
#> Vcells 1447054 11.1    8388608 64.0      32768  2601786 19.9
file.exists(path)
#> [1] FALSE
```

<sup>Created on 2022-08-09 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

File: R/check.R [hunk 4]
```diff
@@ -39,17 +40,18 @@
...
-#'   compatibility. `args = "--output=/foo/bar"` can be used to change the
-#'   check directory.
 #' @param ... Additional arguments passed on to [pkgbuild::build()].
 #' @param vignettes If `FALSE`, do not build or check vignettes, equivalent to
 #'   using `args = '--ignore-vignettes' and `build_args = '--no-build-vignettes'.
-#' @param cleanup Deprecated.
+#' @param cleanup `r lifecycle::badge("deprecated")` `check()` no longer does
+#'   any cleanup, so if you want to inspect the built package, store the return
+#'   value of `check()` and use the `check_dir` argument to leave it in your
+#'   preferred location.
```
Comment (hadley) [inline R/check.R:54]:
```suggestion
#' @param cleanup `r lifecycle::badge("deprecated")` See `check_dir` for details.
```

**Review by hadley — APPROVED** [review]

File: R/check.R (hunk 1 above)
Comment (fweber144) [reply R/check.R:43]: Sorry, I didn't realize I could access the path using `out$checkdir`. For my example above, I was inspecting the temporary folders manually and didn't watch out for the `file<...>` directories. And I forgot that `rm(out)` step before calling `gc()`. So yes, the deletion upon garbage collection (after assigning the return value to some object) seems to work. But assigning the return value to some object doesn't seem to be necessary to keep the `.Rcheck` directory: When running `rcmdcheck::rcmdcheck()` and then going (manually) to the path given in `* using log directory ‘<path_here>’` (`/private/tmp/RtmpK2EwwH/filed4606dfeb7b2/waldo.Rcheck` in your output), the `.Rcheck` folder (and the `<packagename_packageversion>.tar.gz` file) still exists (even after calling `gc()`, but I guess that should not make any difference anyway because the result was not assigned to an object).

So, in summary, this means that my comment [below](https://github.com/r-lib/devtools/pull/2455#discussion_r940917635) was correct but that I need to update r-lib/rcmdcheck#185.

Comment (hadley) [reply R/check.R:43]: @fweber144 if you want me to convince me of that, you'll need to provide a reprex. My reading of the code suggests that it will be cleanup automatically if you don't save it.  (Note that `.Last.value` is likely to hold on to the object for a little bit).

Comment (fweber144) [reply R/check.R:43]: Well, I'm happy with your suggestion [below](https://github.com/r-lib/devtools/pull/2455#discussion_r941289846) and I don't want to change anything here, so I don't really see what I should convince you of.

If it's the statement

> But assigning the return value to some object doesn't seem to be necessary to keep the .Rcheck directory: [...]

I could well imagine that only `.Last.value` causes the `.Rcheck` folder to be kept, but I was just referring to the state right after the check ran through. I don't doubt the deletion upon garbage collection (I believe you!). That's also why I said above that I would need to update https://github.com/r-lib/rcmdcheck/pull/185 so that my former commit https://github.com/fweber144/rcmdcheck/commit/ac7a0473a2b695398fe9d1471b6f1e452ac978e4 is fixed (I would have simply reverted it, but you suggested commit https://github.com/r-lib/rcmdcheck/pull/185/commits/6eeca544305444bd2307ef73907517dff98d06d9 which was fine for me, too).

Comment (fweber144) [reply R/check.R:43]: The problem with a reprex is how to access the `.Rcheck` path by R code if the return value of `rcmdcheck()` is not assigned to some object. But I'll try it anyway.

Comment (hadley) [reply R/check.R:43]: I'd try something like storing `temp_files <- dir(tempdir())` before running, then use `setdiff(dir(tempdir(), temp_files)` afterward.

Comment (fweber144) [reply R/check.R:43]: Good idea, but unfortunately, I just finished my own one and I'm currently lacking the time to pursue further on this.

So I'll try it with a "non-manual" and a "manual" reprex. First, the "non-manual" reprex:

1. Follow steps 1 and 2 from https://github.com/r-lib/devtools/issues/2448#issuecomment-1192858045.
2. While in the folder for that newly created package, run:
    ```r
    con_out <- capture.output(rcmdcheck::rcmdcheck())
    ### Because of `con_out`, the following gc() call would cause the .Rcheck folder
    ### to be deleted (see the "manual" reprex for a demonstration that this gc()
    ### call would not cause the .Rcheck folder to be deleted if
    ### `con_out <- capture.output(...)` wasn't there):
    # gc()
    ###
    pth <- grep("using log directory", con_out, value = TRUE)
    pth <- sub(".*‘(.*)’", "\\1", pth)
    file.exists(pth)
    ## --> TRUE
    gc()
    file.exists(pth)
    ## --> FALSE
    ```

So this probably confirms the `.Last.value` explanation you gave (because `con_out` and `pth` are created after the check).

Now the "manual" reprex where you have to insert the path manually:

1. Follow steps 1 and 2 from https://github.com/r-lib/devtools/issues/2448#issuecomment-1192858045.
2. While in the folder for that newly created package, run:
    ```r
    rcmdcheck::rcmdcheck()
    gc()
    file.exists("<insert .Rcheck path from the 'using log directory' console output>")
    ## --> TRUE
    some_obj <- 1
    file.exists("<insert .Rcheck path from the 'using log directory' console output>")
    ## --> TRUE
    gc()
    file.exists("<insert .Rcheck path from the 'using log directory' console output>")
    ## --> FALSE
    ```

So again, this probably confirms the `.Last.value` explanation you gave.

With these reprexes, I hope you can reproduce what I experienced. But if not, then that's fine for me, too. As I said above, there's nothing I want to change here and there's nothing I wanted to convince you of.
